src/Microsoft.AspNetCore.Routing/Matching/DfaMatcherBuilder.cs

```diff
@@ -145,72 +147,80 @@ public DfaNode BuildDfaTree(bool includeLabel = false)
                         var part = segment.Parts[0];
                         if (segment.IsSimple && part is RoutePatternLiteralPart literalPart)
                         {
-                            DfaNode next = null;
-                            var literal = literalPart.Content;
-                            if (parent.Literals == null ||
-                                !parent.Literals.TryGetValue(literal, out next))
-                            {
-                                next = new DfaNode()
-                                {
-                                    PathDepth = parent.PathDepth + 1,
-                                    Label = includeLabel ? parent.Label + literal + "/" : null,
-                                };
-                                parent.AddLiteral(literal, next);
-                            }
-
-                            nextParents.Add(next);
+                            AddLiteralNode(includeLabel, nextParents, parent, literalPart.Content);
```

Comment:
If you're going to start moving this code around, I'd expect the same treatment for the catch all part (L156-186) - basically either make this low level and awesome, or make it readable and self-documenting

Reply:
I extracted this piece it to a method because it is called from multiple places: for either a literal segment or a parameter with required value segment.